src/udf/src/wasm.rs

```diff
@@ -0,0 +1,305 @@
+
+    pub fn get_or_create() -> Self {
+        use std::sync::LazyLock;
+        static WASM_ENGINE: LazyLock<WasmEngine> = LazyLock::new(WasmEngine::new);
+        WASM_ENGINE.clone()
+    }
+
+    pub async fn compile_and_upload_component(
+        &self,
+        binary: Vec<u8>,
+        wasm_storage_url: &str,
+        identifier: &str,
+    ) -> WasmUdfResult<()> {
+        let object_store = get_wasm_storage(wasm_storage_url).await?;
+        let binary: Bytes = binary.into();
+        object_store
+            .upload(&raw_path(identifier), binary.clone())
+            .await?;
+
+        // This is expensive.
+        let component = Component::from_binary(&self.engine, &binary[..])?;
+        tracing::info!("wasm component loaded");
+
+        // This function is similar to the Engine::precompile_module method where it produces an
+        // artifact of Wasmtime which is suitable to later pass into Module::deserialize. If a
+        // module is never instantiated then it’s recommended to use Engine::precompile_module
+        // instead of this method, but if a module is both instantiated and serialized then this
+        // method can be useful to get the serialized version without compiling twice.
+        let serialized = component.serialize()?;
+        debug!(
+            "compile component, size: {} -> {}",
+            binary.len(),
+            serialized.len()
+        );
+
+        // check the component can be instantiated
+        let mut linker = Linker::new(&self.engine);
+        // A Store is intended to be a short-lived object in a program. No form of GC is
+        // implemented at this time so once an instance is created within a Store it will not be
+        // deallocated until the Store itself is dropped. This makes Store unsuitable for
+        // creating an unbounded number of instances in it because Store will never release this
+        // memory. It's recommended to have a Store correspond roughly to the lifetime of a
+        // "main instance" that an embedding is interested in executing.
+
+        // So creating a Store is cheap?
+
+        let mut store = Store::new(&self.engine, WasmState::try_new()?);
+        wasmtime_wasi::preview2::wasi::command::add_to_linker(&mut linker)?;
+        let (_bindings, _instance) =
+            component::Udf::instantiate_async(&mut store, &component, &linker).await?;
+
+        object_store
+            .upload(&compiled_path(identifier), serialized.into())
+            .await?;
+
+        tracing::debug!(
+            path = compiled_path(identifier),
+            "wasm component compiled and uploaded",
+        );
+
+        Ok(())
+    }
+
+    pub async fn load_component(
+        &self,
+        wasm_storage_url: &str,
+        identifier: &str,
+    ) -> WasmUdfResult<InstantiatedComponent> {
+        let object_store = get_wasm_storage(wasm_storage_url).await?;
+        let serialized_component = object_store.read(&compiled_path(identifier), None).await?;
```

Comment:
FIXME: This stuck forever when I use S3. WHY??? 🥵

Reply:
update: this was because of using `futures` rt inside tokio `block_on` (or sth). Fixed by `tokio::task::block_in_place(|| { tokio::runtime::Handle::current().block_on`